### DISCUSSION TITLE
Dynamic pane management

### DISCUSSION THREAD

**Discussion by cyruseuros** [zellij-org/zellij#2409] [Ideas]
I realize this is somewhat of a niche request and zellij has been serving me quite well as-is recently, however I don't think it would be very large in scope, and there is interest out there (https://github.com/martanne/dvtm/issues/10#issuecomment-1141451337).

From what I can tell all that would be necessary to implement a full dwm/dvtm emulation mode is:
- Make `MoveFocus "Up"`/`MoveFocus "Down"` loop around or implement `MoveFocus "Next"`/`MoveFocus "Previous"`
- implement the concept of "master" panes
- implement a dwm/dvtm mode just like the tmux mode (dvtm only has 13 shortcuts and they cover almost every use case very elegantly

The tagging feature for tabs would also be nice, but #1750 would need to be implemented first.

**Comment by imsnif** [comment]
hey @cyruseuros - I personally do not know dvtm and have never used it. Could you please explain in a little bit more detail what you're after? Sort of a run-down for people like me who have no reference?

**Comment by cyruseuros** [comment]
Sure thing, dvtm is inspired by [dwm](https://dwm.suckless.org/). It's a *dynamic* window manager (akin to zellij swap layouts) as opposed to a manual window manager (i.e. you don't have to say things like split left/right/bottom, the layout has that logic built-in). To quote the docs:
> Windows are managed in a master and stacking area. The master area contains the window which currently needs most attention, whereas the stacking area contains all other windows.
Windows can be popped from the stack (at any index) to the master area, or pushed back to the stack

The zellij stacking layout (minimizing all windows but the "main" one on the stack is a common tweak introduced by tiling WMs that implement this layout.

Another unique feature is that
> Windows are grouped by tags [not tabs though they look like it]. Each window can be tagged with one or multiple tags. Selecting certain tags displays all windows with these tags.

Meaning that you can have the same window in multiple tabs. It's a pretty flexible model that reflects the internal algo, and it would make implementing #1750 a breeze.

Because the tiling algo focuses on what you want done (focus on something, have a reference thing, make focus thing more prominent) as opposed to how you want it done (split window left/right etc.) the number of commands you need for everyday use is very small, so being limited to a single prefix key doesn't actually feel limiting (unlike, say tmux) - which is why I think it'll be more than fine as a single mode.

**Comment by imsnif** [comment]
Thanks for the explanation @cyruseuros - I really like this approach. In fact, somewhere in the back of my mind I am considering similar approaches as a way to solve our (out of the box) colliding keybindings problem.

I think the way I'd like to look at this now is to implement all the small features (eg. there's an open PR for looping focus in all directions, and I hope to get time to work on moving a pane to a new tab soon) and allow users to configure this sort of behavior for themselves rather than including it as a separate mode.

The only thing I'm not clear about here is the "master" pane behavior you mentioned. How is it set? Dynamically at runtime, or?

**Reply by cyruseuros** [reply]
The master pane is the last one that is created (though most dwm-likes also provide the option to create at the top of the stack). You can swap buffers between panes in an arbitrary fashion but the only thing you really end up doing in practice is moving a buffer from a stack pane of particular interest to the master pane. The only resizing you're also likely to do is growing/shrinking the master pane.

Here's a demo of the flow: ![demo](https://raw.githubusercontent.com/martanne/dvtm/gh-pages/screencast.gif#center) 

Here's the manpage that illustrates just how simple the keybindings are: https://linux.die.net/man/1/dvtm

**Reply by cyruseuros** [reply]
I should mention that although there is still internally a master pane (if I'm not mistaken) there is also a master and stack area - you can increase/decrease the number of panes in the master area, but one of them - typically at index 0 - will still be the master pane for the purposes of window creation for example.

**Reply by imsnif** [reply]
Thanks for the demo and detailed explanation! My main issue with this or similar setups is that I feel they are very hard to communicate to the user. It's something we as users need to "learn" rather than just use out of the box.

For this reason I'm going to work on implementing features that will facilitate creating such a setup in our plugin system. I'm doing work in that area these days anyway.

**Reply by cyruseuros** [reply]
@imsnif appreciate you hearing me out - fair comment and very much valid conclusion. It's actually kind of true for all auto-tilers like bspwm, dwm, Xmonad (by default)... 

That being said I think there are two benefits for zellij to not relegating this to the plugin system:
1. Adoption: In the tiling window manager world, the way new players like Qtile acquired a substantial userbase was by having the ability to emulate *several* alternatives out of the box (and do so well) not just in terms of keybindings (which I think are less important, muscle memory can be built up quickly enough) but fundamental concepts and layouts. The master-stack layout in particular will be familiar to anyone that used any of the big tiling WMs (AwesomeWM, Xmonad, dwm) - it's only really new in the mux space.
2. Consistency/Discoverability: There is so much zellij does - and I suspect I only know 20% of it, let alone the keybinds for it (1 bar of auto suggestions helps, but not a lot). Event the default automatic layout has so many rules that after months of using it I still occasionally ask myself what happened to my panes, and the keybinding clashes have so many issues referencing them #146 #2073, #2193, #1399. The master-stack layout does take 5 min to grok, however once you do that's all you need to know - I would argue that's far less time than what's needed to get familiar with zellij now.

If it's architecturally cleaner to implement as a plugin there's no reason not to do it, though I would advise the plugin being shipped with the mux by default (people don't really take that much time to test a new program out and if it ships with something they're familiar with they're far more likely to stay)

All else aside, I'm just some dude that's immensely grateful for your work on an awesome, powerful piece of software. It's 100% your call to make and I'll be happy with zellij either way (seriously, it's *the* thing that finally made me comfortable in the terminal), these are just my 2 cents.

**Comment by cyruseuros** [comment]
Side question, are there tiny things I could help with in this endeavor (or whatever alternative approach you choose to take)? I've only writen toy programs in rust so I'll probably be slow, but I'm happy to chip in any way I can.
